crates/extrinsics/src/call.rs

```diff
@@ -104,151 +212,90 @@ impl CallCommand {
             .map(|bv| bv.denominate_balance(&token_metadata))
             .transpose()?;
         let call_request = CallRequest {
-            origin: account_id(signer),
+            origin: account_id(&self.signer),
             dest: self.contract.clone(),
             value: self.value.denominate_balance(&token_metadata)?,
             gas_limit: None,
             storage_deposit_limit,
-            input_data,
+            input_data: self.call_data.clone(),
         };
         state_call(&url, "ContractsApi_call", call_request).await
     }
 
-    async fn call(
-        &self,
-        client: &Client,
-        data: Vec<u8>,
-        signer: &Keypair,
-        transcoder: &ContractMessageTranscoder,
-    ) -> Result<(), ErrorVariant> {
+    pub async fn call(&self, gas_limit: Weight) -> Result<DisplayEvents, ErrorVariant> {
         tracing::debug!("calling contract {:?}", self.contract);
-
-        let gas_limit = self
-            .pre_submit_dry_run_gas_estimate(client, data.clone(), signer)
-            .await?;
-
-        if !self.extrinsic_opts.skip_confirm {
-            prompt_confirm_tx(|| {
-                name_value_println!("Message", self.message, DEFAULT_KEY_COL_WIDTH);
-                name_value_println!("Args", self.args.join(" "), DEFAULT_KEY_COL_WIDTH);
-                name_value_println!(
-                    "Gas limit",
-                    gas_limit.to_string(),
-                    DEFAULT_KEY_COL_WIDTH
-                );
-            })?;
-        }
-
-        let token_metadata = TokenMetadata::query(client).await?;
+        let token_metadata = TokenMetadata::query(&self.client).await?;
 
         let call = api::tx().contracts().call(
             self.contract.clone().into(),
             self.value.denominate_balance(&token_metadata)?,
             gas_limit.into(),
-            self.extrinsic_opts.storage_deposit_limit(&token_metadata)?,
-            data,
+            self.opts.storage_deposit_limit(&token_metadata)?,
+            self.call_data.clone(),
         );
 
-        let result = submit_extrinsic(client, &call, signer).await?;
+        let result = submit_extrinsic(&self.client, &call, &self.signer).await?;
 
-        let display_events =
-            DisplayEvents::from_events(&result, Some(transcoder), &client.metadata())?;
+        let display_events = DisplayEvents::from_events(
+            &result,
+            Some(&self.transcoder),
+            &self.client.metadata(),
+        )?;
 
-        let output = if self.output_json {
-            display_events.to_json()?
-        } else {
-            display_events
-                .display_events(self.extrinsic_opts.verbosity()?, &token_metadata)?
-        };
-        println!("{output}");
-
-        Ok(())
+        Ok(display_events)
     }
 
     /// Dry run the call before tx submission. Returns the gas required estimate.
-    async fn pre_submit_dry_run_gas_estimate(
+    pub async fn pre_submit_dry_run_gas_estimate(
```

Comment:
I'm wondering whether a simpler name can be used, e.g. something like `estimate_gas`?